### PR TITLE
feat: core/drift.ts + core/differ.ts — three-way merge engine (#11)

### LINKED ISSUE #11: core/drift.ts + core/differ.ts — three-way merge engine
## Spec Reference
`docs/IMPLEMENTATION.md` §4.8 (drift), §4.9 (differ), §17.4 (merge impl), §17.5 (edge cases)

## Tasks
- drift.ts: detectDrift() → DriftReport (managed, modified, volatile, missing, orphaned)
- differ.ts: computeMergeAction() → MergeAction (skip, overwrite, auto_merge, conflict)
- differ.ts: threeWayMerge() via node-diff3 (Khanna-Myers)
- Conflict markers: <<<<<<< yours / ======= / >>>>>>> shard update
- ShardMindError with typed codes for merge failures

## Acceptance
- All 17 fixture scenarios pass
- Edge case fixtures pass (frontmatter merge, empty file, hash-identical, encoding)
- Merge stats (linesUnchanged, linesAutoMerged, linesConflicted) are accurate

### PR DESCRIPTION
**Pull request by breferrari**
Closes #11. Milestone 3 of the ROADMAP — the update engine that makes ShardMind different from every other "stamp-out-a-template" tool.

## Summary

- **`source/core/drift.ts`** — `detectDrift()` classifies every tracked file into `managed / modified / volatile / missing / orphaned` by sha256-comparing disk content against `state.json.rendered_hash`. Non-recursive orphan scan: parent directory of any tracked file = tracked directory; files in a tracked dir not in `state.files` = orphans. Engine scaffolding (`shard-values.yaml`, `.shardmind/`) and third-party metadata (`.git/`, `.obsidian/`) are excluded at the source. Reads are bounded by `mapConcurrent(items, 32, fn)` so large vaults can't exhaust file descriptors.
- **`source/core/differ.ts`** — `computeMergeAction()` per spec §4.9: skip when `base === ours`, overwrite silently for managed files, three-way merge for modified files. `threeWayMerge()` uses `node-diff3`'s `diff3MergeRegions` so stable regions with `buffer === 'a'|'b'` can be classified as auto-merged vs `'o'` unchanged, which flat `diff3Merge` can't discriminate. A `LineInterner` maps each unique line to an integer-named token before diff3 sees it — this sidesteps a latent upstream crash where lines matching `Object.prototype` member names (`constructor`, `__proto__`, `toString`, …) collide with the library's internal `{}` hash map. Filed upstream as [bhousel/node-diff3#86](https://github.com/bhousel/node-diff3/issues/86) with fix PR [#87](https://github.com/bhousel/node-diff3/pull/87); removal of our workaround tracked in #49.
- **`source/core/renderer.ts`** — exposes `renderString()` for in-memory template rendering (no disk I/O) via a lazy isolated `nunjucks.Environment`. `renderFrontmatterSafely()` auto-recovers when naive render produces YAML-hostile output (e.g. `owner: {{ name }}` with `name = "Alice: AI researcher"`) by re-rendering with every string leaf JSON-encoded; non-string values keep their YAML type. A `WeakMap` seen-set guards the walker against circular references in values.
- **`source/runtime/errors.ts`** — typed `ErrorCode` string union (39 codes, grouped by domain). `ShardMindError.code` is now strictly typed; typos surface at compile time. Exported from `shardmind/runtime` for hook consumers.
- **`source/runtime/errno.ts`** — shared `errnoCode` / `isEnoent` with runtime `typeof === 'string'` guard. Collapsed 8 copies of the `err instanceof Error && 'code' in err ? ...` pattern across core + runtime.
- **CRLF tolerance** — `threeWayMerge` splits on `/\r?\n/`, so a user's Windows-saved file (CRLF `theirs`) doesn't fight LF `base`/`ours` rendered by the engine. Merged output is always LF.
- **Cross-OS CI matrix** — `{ubuntu, windows, macos} × {node 22, node 24}` with `fail-fast: false` and `defaults.run.shell: bash`. Windows caught the CRLF regression.
- Conflict markers use git vocabulary: `<<<<<<< yours` / `=======` / `>>>>>>> shard update`.

## Adversarial hardening

Deliberately pushed the engine against 16 attack categories. Four real bugs surfaced and were fixed:

1. `node-diff3` crash on `Object.prototype` line content — fixed via `LineInterner` (and filed upstream).
2. YAML-hostile frontmatter values crashing render — fixed via auto-recovery.
3. Circular references in values stack-overflowing the recovery walker — fixed with `WeakMap` seen-set.
4. Stats over-counted by 1 on newline-terminated content — fixed by subtracting the trailing-empty token post-loop.

Twelve other categories held up: line endings (CRLF/LF/CR/mixed/no-trailing), conflict-marker injection (user content containing the marker strings), unicode/BOM/null bytes, size extremes (10K chars, 5K lines, many conflicts), idempotence & convergence, conflict boundaries (start/end/whole-file), drift races, exotic value types (Date/Function/Symbol), drift performance (500 files in <5s), token interning stress, whitespace-sensitive diffs, stats bookkeeping.

## Test plan

- [x] `npm test` — **300 tests pass** across 26 files
- [x] `npm run typecheck` — strict mode clean, zero `any` / `as any` / `@ts-ignore`
- [x] `npm run build` — runtime bundle 7.66 KB (well under the 30 KB target)
- [x] 20 fixture scenarios (17 spec + 3 edge cases: empty file, UTF-8 non-ASCII, frontmatter merge on modified ownership)
- [x] 12 direct `detectDrift` unit tests covering every `DriftReport` bucket including 5 orphan scenarios
- [x] 7 direct `threeWayMerge` unit tests pinning down stats accounting (caught a bug the fixture suite missed)
- [x] 6 fast-check property invariants × 200 cases = **1 200 generative scenarios** (caught the `node-diff3` prototype crash)
- [x] 60+ adversarial stress tests across 16 attack categories
- [x] 2 CRLF-robustness tests
- [x] Cross-OS CI matrix green on all 6 jobs
- [x] ROADMAP.md Milestone 3 fully checked; issue #47 (orphan detection) pulled forward and closed
- [x] Spec docs (`docs/ARCHITECTURE.md` §17, `docs/IMPLEMENTATION.md` §4.8/§4.9) updated to match implementation
- [x] `CHANGELOG.md` [Unreleased] section populated
- [x] Two rounds of Copilot review addressed with disposition + commit refs on every thread

## Tracking

- Upstream workaround removal: #49 (drop `LineInterner` once bhousel/node-diff3#87 ships).
- v0.2 Engine polish bullet in ROADMAP mirrors #49.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: source/core/differ.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+ * `>>>>>>> shard update` wraps ours).
+ */
+export function threeWayMerge(
+  base: string,
+  theirs: string,
+  ours: string,
+): ThreeWayMergeResult {
+  const baseLines = base.split('\n');
+  const theirsLines = theirs.split('\n');
+  const oursLines = ours.split('\n');
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/differ.ts]: Good catch — valid concern, fixed in 31c9e04.

`threeWayMerge` now splits all three inputs on `/\r?\n/`, so a CRLF `theirs` on Windows no longer fights LF `base` / `ours` (renderer output is always LF). Merged output stays LF; a comment on the split documents the choice and flags that callers should convert to platform-native line endings at the write boundary if ever needed.

Covered by `tests/unit/differ-line-endings.test.ts` — one test for identity (CRLF `theirs` matching LF `base`/`ours` → no conflict) and one for a genuine conflict that only CRLF-normalization could reveal.

File: source/core/drift.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  const trackedPaths = new Set(Object.keys(state.files));
+
+  const [classified, orphaned] = await Promise.all([
+    Promise.all(
+      Object.entries(state.files).map(([relPath, file]) =>
+        classifyFile(vaultRoot, relPath, file),
+      ),
+    ),
+    detectOrphans(vaultRoot, trackedPaths),
+  ]);
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/drift.ts:72]: EMFILE risk is real at scale. Fixed in df54469: added a `mapConcurrent(items, 32, fn)` helper that caps in-flight file handles at 32 while preserving input order. 32 stays well under typical OS fd limits (macOS defaults to ~256) while still saturating disk on realistic vaults.

File: source/core/drift.ts [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  relPath: string,
+  file: FileState,
+): Promise<Classified> {
+  // FileState.ownership uses the literal 'user' for volatile files (what the
+  // install engine writes to state.json); DriftEntry.ownership reports it as
+  // 'volatile' because that is the reporting-layer vocabulary used by the
+  // update command. Intentional naming gap, not a bug.
+  if (file.ownership === 'user') {
+    return { bucket: 'volatile', entry: volatileEntry(relPath, file) };
+  }
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/drift.ts:98]: Already addressed — the 'user' → 'volatile' naming gap is documented both in the inline comment at drift.ts:72-75 and in `docs/IMPLEMENTATION.md` §4.8 (updated in cabe217). Leaving both names as-is: 'user' in the persisted state because that's what the install engine writes; 'volatile' in the reporting layer because that's what the update command displays. Renaming either breaks backwards compatibility with existing state.json files.

File: source/core/drift.ts [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    const dir = path.posix.dirname(toPosixPath(relPath));
+    trackedDirs.add(dir === '.' ? '' : dir);
+  }
+
+  const perDir = await Promise.all(
+    [...trackedDirs].map(relDir => scanDirForOrphans(vaultRoot, relDir, trackedPaths)),
+  );
+
+  return perDir.flat().sort();
+}
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/drift.ts:179]: Stale — orphan detection was pulled into v0.1 during review and the scope was reconciled. PR body was refreshed, ROADMAP.md Milestone 3 bullet tracks the v0.1 landing, and issue #47 is closed with the rationale (e5304dd). `detectDrift` now returns a populated `orphaned` array, which matches what the tests exercise.

File: source/core/differ.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  // Normalize line endings to LF before splitting. `base` and `ours` are
+  // renderer output (always LF), but `theirs` comes from disk and may be
+  // CRLF on Windows. Without this, every line in theirs would trail with
+  // '\r', producing spurious conflicts. Merged output is LF; callers that
+  // need platform-native line endings convert at the write boundary.
+  const regions: IRegion<string>[] = diff3MergeRegions(
+    theirs.split(/\r?\n/),
+    base.split(/\r?\n/),
+    ours.split(/\r?\n/),
+  );
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/differ.ts]: Good catch on the stats inflation. Fixed in df54469 — when all three inputs end with `\n`, `split(/\r?\n/)` produces a trailing "" that diff3 emits as a stable region and inflates `linesUnchanged` by 1. After the merge loop, we now detect this and subtract one from `linesUnchanged`.

Kept the trailing "" flowing through diff3 itself rather than stripping it — a newline-status difference between theirs and ours is a legitimate auto-merge concern, not a pre-merge normalization. My first attempt stripped + re-appended based on ours' style, which broke the property test "theirs-only-changed ⇒ output equals theirs". The stats-only correction preserves merge semantics and fixes the off-by-one in reporting.

File: source/core/drift.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    throw err;
+  }
+
+  const orphans: string[] = [];
+  for (const entry of entries) {
+    if (!entry.isFile()) continue;
+    if (UNSCANNED_DIRS.has(entry.name)) continue;
+    const rel = relDir ? `${relDir}/${entry.name}` : entry.name;
+    if (trackedPaths.has(rel)) continue;
+    if (ENGINE_RESERVED_FILES.has(rel)) continue;
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/drift.ts:200]: Correct — the `UNSCANNED_DIRS.has(entry.name)` check ran after `entry.isFile()` filtered out directories, so it was effectively dead code.

Fixed in df54469 by moving the check to `detectOrphans` where `trackedDirs` is derived: any directory whose top segment is `.shardmind`, `.git`, or `.obsidian` is dropped from the set before any `readdir` is issued. If a shard somehow tracks `.git/HEAD`, that whole directory is skipped at the source. Renamed the constant to `UNSCANNED_DIR_NAMES` to reflect the new role.

File: source/core/renderer.ts [hunk 7]
```diff
@@ -14,12 +14,43 @@ import { ShardMindError } from '../runtime/types.js';
...
+/**
+ * Isolated env for rendering a template from a string (no filesystem loader).
+ * Lazily constructed so the `nunjucks.Environment` is only built when needed
+ * and never pollutes the module's global `nunjucks.configure()` state.
+ */
+let defaultStringEnv: nunjucks.Environment | undefined;
+
+function getDefaultStringEnv(): nunjucks.Environment {
+  if (!defaultStringEnv) {
+    defaultStringEnv = new nunjucks.Environment(null, NUNJUCKS_OPTS);
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/renderer.ts]: Fair — swapped to `new nunjucks.Environment([], NUNJUCKS_OPTS)` in df54469. Added a comment noting the empty loader array explicitly represents 'no filesystem resolution', which is the intent for in-memory string rendering.

File: tests/unit/drift.test.ts [hunk 8]
```diff
@@ -92,12 +108,13 @@ async function loadFiles(dir: string): Promise<FixtureFiles> {
...
+  // expected-output.md is optional — conflict scenarios omit it because
+  // the marker format is implementation-defined.
   let expectedOutput: string | null = null;
   try {
     expectedOutput = await fsp.readFile(path.join(base, 'expected-output.md'), 'utf-8');
   } catch {
-    // Some scenarios (e.g. conflicts) deliberately omit expected-output.md
-    // because the merge markers depend on implementation details.
+    /* no-op */
   }
```
[the remark this answers is not in the document]
Comment (breferrari) [reply tests/unit/drift.test.ts:121]: Right — the blanket catch was swallowing permissions/path errors too. Narrowed to ENOENT in df54469; any other read failure now throws so fixture-setup problems surface loudly.

File: source/runtime/errno.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+ * `err instanceof Error && 'code' in err ? (err as NodeJS.ErrnoException).code : undefined`
+ * dance doesn't recur in every file that touches the filesystem.
+ *
+ * Lives in runtime/ so both runtime modules (for hook scripts) and core
+ * modules can import without crossing the one-way runtime → core boundary.
+ */
+
+export function errnoCode(err: unknown): string | undefined {
+  if (err !== null && typeof err === 'object' && 'code' in err) {
+    return (err as { code?: string }).code;
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/runtime/errno.ts]: Good defensive catch. Fixed in df54469 with the suggested `typeof === 'string'` guard. Node's types document `code` as string, but a non-string on a custom thrown object would otherwise silently violate our return type.

File: source/core/drift.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  const trackedPaths = new Set(Object.keys(state.files));
+
+  const [classified, orphaned] = await Promise.all([
+    Promise.all(
+      Object.entries(state.files).map(([relPath, file]) =>
+        classifyFile(vaultRoot, relPath, file),
+      ),
+    ),
+    detectOrphans(vaultRoot, trackedPaths),
+  ]);
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/drift.ts]: Fixed in df54469 — `trackedPaths` is now normalized to forward slashes on construction, so the `trackedPaths.has(rel)` check inside `scanDirForOrphans` works whether state.files was written with posix or native separators. `install-executor` already emits posix, but the defensive normalization costs nothing and removes a silent-failure mode.

File: source/core/differ.ts [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+ * the git conflict-marker vocabulary (`<<<<<<< yours` wraps theirs,
+ * `>>>>>>> shard update` wraps ours).
+ */
+export function threeWayMerge(
+  base: string,
+  theirs: string,
+  ours: string,
+): ThreeWayMergeResult {
+  // Intern every unique line to an integer-named token. See the block
+  // comment on LINE_SPLIT for why this is load-bearing.
```
[the remark this answers is not in the document]
Comment (breferrari) [reply source/core/differ.ts]: Stale reference fixed in df54469. Updated the comment to point at the LineInterner/prototype-collision rationale (which is the load-bearing part); `LINE_SPLIT` is just about CRLF tolerance.

File: tests/unit/merge-adversarial.test.ts [hunk 12]
```diff
@@ -0,0 +1,701 @@
...
+  install_date: '2026-04-19',
+  year: '2026',
+};
+
+describe('merge adversarial — sentinel corruption vectors', () => {
+  // differ.ts prefixes every line with U+0001 to sidestep the node-diff3
+  // prototype-pollution crash. If the strip pass is implemented naively
+  // (global replace), legitimate user content containing U+0001 would be
+  // mangled. Probe for it.
+
```
[the remark this answers is not in the document]
Comment (breferrari) [reply tests/unit/merge-adversarial.test.ts]: Rewrote the describe block in df54469 to describe the current `LineInterner`-based approach. Kept the round-trip tests as a regression guard — they prove user content containing `U+0001` survives the merge, which an earlier sentinel-prefix implementation would have mangled. Also dropped two `.not.toContain('\u0001')` assertions elsewhere that were sentinel-era leftovers.
